### PR TITLE
add yolo_box for detection box calc in YOLOv3

### PR DESCRIPTION
**Pull request by heavengate**

### CODE REVIEW

File: paddle/fluid/operators/detection/yolo_box_op.cc [hunk 1]
```diff
@@ -0,0 +1,157 @@
+/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserve.
```
Comment (jerrywgz) [inline paddle/fluid/operators/detection/yolo_box_op.cc]: 2018->2019

File: paddle/fluid/operators/detection/yolo_box_op.cc [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+                 "input, so 32, 16, 8 should be set for the first, second, "
+                 "and thrid YoloBox operators.")
+        .SetDefault(32);
+    AddAttr<float>("conf_thresh",
+                   "The confidence scores threshold of detection boxes."
+                   "boxes with confidence scores under threshold should"
+                   "be ignored.")
+        .SetDefault(0.01);
+    AddComment(R"DOC(
+         This operator generate YOLO detection boxes fron output of YOLOv3 network.
```
Comment (jerrywgz) [inline paddle/fluid/operators/detection/yolo_box_op.cc]: fron->from

File: paddle/fluid/operators/detection/yolo_box_op.cc [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+         
+         The output of previous network is in shape [N, C, H, W], while H and W
+         should be the same, specify the grid size, each grid point predict given
+         number boxes, this given number is specified by anchors, it should be 
+         half anchors length, which following will be represented as S. In the 
+         second dimention(the channel dimention), C should be S * (class_num + 5),
```
Comment (jerrywgz) [inline paddle/fluid/operators/detection/yolo_box_op.cc]: dimension

File: paddle/fluid/operators/detection/yolo_box_op.cc [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+         b_h = p_h e^{t_h}
+         $$
+
+         While $$c_x, c_y$$ is the left top corner of current grid and $$p_w, p_h$$
+         is specified by anchors.
+
+         The logistic scores of the 5rd channel of each anchor prediction boxes
+         represent the confidence score of each prediction scores, and the logistic
+         scores of the last class_num channels of each anchor prediction boxes 
+         represent the classifcation scores. Boxes with confidence scores less then
```
Comment (jerrywgz) [inline paddle/fluid/operators/detection/yolo_box_op.cc]: than

File: paddle/fluid/operators/detection/yolo_box_op.cc [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+         conf_thresh should be ignored, and boxes final scores if the products result
```
Comment (jerrywgz) [inline paddle/fluid/operators/detection/yolo_box_op.cc]: last sentence is confused

File: paddle/fluid/operators/detection/yolo_box_op.cc (hunk 1 above)
Comment (heavengate) [reply paddle/fluid/operators/detection/yolo_box_op.cc]: Done

File: paddle/fluid/operators/detection/yolo_box_op.cc (hunk 2 above)
Comment (heavengate) [reply paddle/fluid/operators/detection/yolo_box_op.cc]: Done

File: paddle/fluid/operators/detection/yolo_box_op.cc (hunk 3 above)
Comment (heavengate) [reply paddle/fluid/operators/detection/yolo_box_op.cc]: Done

File: paddle/fluid/operators/detection/yolo_box_op.cc (hunk 4 above)
Comment (heavengate) [reply paddle/fluid/operators/detection/yolo_box_op.cc]: Done

File: paddle/fluid/operators/detection/yolo_box_op.cc (hunk 5 above)
Comment (heavengate) [reply paddle/fluid/operators/detection/yolo_box_op.cc]: Done

**Comment by heavengate** [comment]
@haowang101779990 please check doc

**Comment by haowang101779990** [comment]
Thanks, here is my review of API :

图片


图片

**Comment by haowang101779990** [comment]
C=S*(class_num+5). ==> C= S*(5+class_num)

for 5 (xyhw, confidence channels)  comes prior to class_num

**Comment by heavengate** [comment]
> C=S*(class_num+5). ==> C= S*(5+class_num)
> 
> for 5 (xyhw, confidence channels) comes prior to class_num

Done

File: paddle/fluid/operators/detection/yolo_box_op.cc [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+                                   ctx.GetPlace());
+  }
+};
+
+class YoloBoxOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("X",
+             "The input tensor of YoloBox operator, "
+             "This is a 4-D tensor with shape of [N, C, H, W]. "
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/yolo_box_op.cc]: , This is -> is, 去掉 , This

**Comment by shanyi15** [comment]
有默认值的参数，请在参数说明处再强调下

**Review by shanyi15 — DISMISSED** [review]
LG comments

File: python/paddle/fluid/layers/detection.py [hunk 7]
```diff
@@ -609,6 +610,83 @@ def yolov3_loss(x,
...
+        TypeError: Attr class_num of yolo box must be an integer
+        TypeError: Attr conf_thresh of yolo box must be a float number
+
+    Examples:
+
+    .. code-block:: python
+
+        x = fluid.layers.data(name='x', shape=[255, 13, 13], dtype='float32')
+        anchors = [10, 13, 16, 30, 33, 23]
+        loss = fluid.layers.yolov3_loss(x=x, class_num=80, anchors=anchors,
```
Comment (qingqing01) [inline python/paddle/fluid/layers/detection.py]: yolov3_loss -> yolov3_box

File: paddle/fluid/operators/detection/yolo_box_op.cc [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+                 "input, so 32, 16, 8 should be set for the first, second, "
+                 "and thrid YoloBox operators.")
+        .SetDefault(32);
+    AddAttr<float>("conf_thresh",
+                   "The confidence scores threshold of detection boxes. "
+                   "Boxes with confidence scores under threshold should "
+                   "be ignored.")
+        .SetDefault(0.01);
+    AddComment(R"DOC(
+         This operator generate YOLO detection boxes from output of YOLOv3 network.
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/yolo_box_op.cc]: This operator generate ->  This operator generates

File: paddle/fluid/operators/detection/yolo_box_op.cu [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    const int w = input->dims()[3];
+    const int box_num = boxes->dims()[1];
+    const int an_num = anchors.size() / 2;
+    int input_size = downsample_ratio * h;
+
+    Tensor anchors_t, cpu_anchors_t;
+    auto cpu_anchors_data =
+        cpu_anchors_t.mutable_data<int>({an_num * 2}, platform::CPUPlace());
+    std::copy(anchors.begin(), anchors.end(), cpu_anchors_data);
+    TensorCopySync(cpu_anchors_t, ctx.GetPlace(), &anchors_t);
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/yolo_box_op.cu]: There is no need to use sync copy, use async code based on stream

File: python/paddle/fluid/tests/unittests/test_yolo_box_op.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/test_yolo_box_op.py]: 2018 -> 2019

File: paddle/fluid/operators/detection/yolo_box_op.cc (hunk 6 above)
Comment (heavengate) [reply paddle/fluid/operators/detection/yolo_box_op.cc]: Done

File: paddle/fluid/operators/detection/yolo_box_op.cc (hunk 8 above)
Comment (heavengate) [reply paddle/fluid/operators/detection/yolo_box_op.cc]: Done

File: python/paddle/fluid/layers/detection.py (hunk 7 above)
Comment (heavengate) [reply python/paddle/fluid/layers/detection.py]: Done

File: paddle/fluid/operators/detection/yolo_box_op.cu (hunk 9 above)
Comment (heavengate) [reply paddle/fluid/operators/detection/yolo_box_op.cu]: Done

**Review by qingqing01 — DISMISSED** [review]

File: python/paddle/fluid/tests/unittests/test_yolo_box_op.py (hunk 10 above)
Comment (heavengate) [reply python/paddle/fluid/tests/unittests/test_yolo_box_op.py]: Done

**Review by shanyi15 — DISMISSED** [review]
LG comments

**Review by qingqing01 — DISMISSED** [review]

**Review by XiaoguangHu01 — DISMISSED** [review]

**Review by shanyi15 — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]

**Pull request merged**